resolver/src/main/java/io/netty/resolver/HostsFileEntriesProvider.java

```diff
@@ -0,0 +1,291 @@
+
+        /**
+         * Parses the hosts file.
+         *
+         * @return a new {@link HostsFileEntriesProvider}
+         * @throws IOException file could not be read
+         */
+        HostsFileEntriesProvider parse() throws IOException;
+
+        /**
+         * Parses the hosts file.
+         *
+         * @return a new {@link HostsFileEntriesProvider}
+         */
+        HostsFileEntriesProvider parseSilently();
+    }
+
+    /**
+     * Creates a parser for {@link HostsFileEntriesProvider}.
+     *
+     * @return a new {@link HostsFileEntriesProvider.Parser}
+     */
+    public static Parser parser() {
+        return new ParserImpl();
+    }
+
+    static final HostsFileEntriesProvider EMPTY =
+            new HostsFileEntriesProvider(
+                    Collections.<String, List<InetAddress>>emptyMap(),
+                    Collections.<String, List<InetAddress>>emptyMap());
+
+    private final Map<String, List<InetAddress>> ipv4Entries;
+    private final Map<String, List<InetAddress>> ipv6Entries;
+
+    HostsFileEntriesProvider(Map<String, List<InetAddress>> ipv4Entries, Map<String, List<InetAddress>> ipv6Entries) {
+        this.ipv4Entries = Collections.unmodifiableMap(new HashMap<String, List<InetAddress>>(ipv4Entries));
+        this.ipv6Entries = Collections.unmodifiableMap(new HashMap<String, List<InetAddress>>(ipv6Entries));
+    }
+
+    /**
+     * The IPv4 entries.
+     *
+     * @return the IPv4 entries
+     */
+    public Map<String, List<InetAddress>> ipv4Entries() {
+        return ipv4Entries;
+    }
+
+    /**
+     * The IPv6 entries.
+     *
+     * @return the IPv6 entries
+     */
+    public Map<String, List<InetAddress>> ipv6Entries() {
+        return ipv6Entries;
+    }
+
+    private static final class ParserImpl implements Parser {
+
+        private static final String WINDOWS_DEFAULT_SYSTEM_ROOT = "C:\\Windows";
+        private static final String WINDOWS_HOSTS_FILE_RELATIVE_PATH = "\\system32\\drivers\\etc\\hosts";
+        private static final String X_PLATFORMS_HOSTS_FILE_PATH = "/etc/hosts";
+
+        private static final Pattern WHITESPACES = Pattern.compile("[ \t]+");
+
+        private static final InternalLogger logger = InternalLoggerFactory.getInstance(Parser.class);
+
+        private Charset[] charsets = new Charset[]{Charset.defaultCharset()};
+        private File file;
+        private Reader reader;
+
+        @Override
+        public Parser charsets(Charset... charsets) {
+            this.charsets = checkNotNull(charsets, "charsets");
+            return this;
+        }
+
+        @Override
+        public Parser file(File file) {
+            this.file = checkNotNull(file, "file");
+            this.reader = null;
+            return this;
+        }
+
+        @Override
+        public Parser reader(Reader reader) {
+            this.reader = checkNotNull(reader, "reader");
+            this.file = null;
+            return this;
+        }
```

Comment:
imho its a bit strange that you can modify the `ParserImpl` and `parse()` depends on the "state". Either make this a proper builder or change `parse(...)` to take arguments (which I think may be the best).

Reply:
fixed

Reply:
Not really fixed... I think both methods above should also not exist.